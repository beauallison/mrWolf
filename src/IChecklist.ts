export default interface IChecklist {
  name: string;
  complete: boolean;
}
