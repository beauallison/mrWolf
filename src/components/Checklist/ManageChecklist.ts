import IChecklist from '../../IChecklist';

export interface IManageItem {
  item?: IChecklist;
  index: number;
}

export interface IManageChecklist {
  checklist?: IChecklist[];
  updateChecklist: Function;
}

export const ManageChecklist = ({ checklist, updateChecklist }: IManageChecklist) => {
  const createItem = (name: String) => {
    const newItem = {
      name,
      complete: false,
    } as IChecklist;

    const newChecklist = [newItem, ...checklist];
    return updateChecklist(newChecklist);
  };

  const clearCompletedItems = () => {
    const newChecklist = checklist.filter(({ complete }) => !complete);
    return updateChecklist(newChecklist);
  };

  const toggleItem = ({ item, index }: IManageItem) => {
    const value = !item.complete;
    const newItem = {
      name: item.name,
      complete: value,
    } as IChecklist;
    const newChecklist = [...checklist];
    newChecklist[index] = newItem;
    return updateChecklist(newChecklist);
  };

  const deleteItem = ({ index }: IManageItem) => {
    const newChecklist = [...checklist];
    newChecklist.splice(index, 1);
    return updateChecklist(newChecklist);
  };

  return {
    createItem,
    toggleItem,
    deleteItem,
    clearCompletedItems,
  };
};
