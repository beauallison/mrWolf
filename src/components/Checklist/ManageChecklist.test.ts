import IChecklist from '../../IChecklist';
import { ManageChecklist } from './ManageChecklist';

const defaultChecklist = [
  {
    name: 'Live by harmless untruths',
    complete: false,
  },
  {
    name: 'Research the foma',
    complete: false,
  },
  {
    name: 'Busy busy busy',
    complete: true,
  },
] as IChecklist[];

describe('ManageChecklist', () => {
  let checklist = defaultChecklist;
  const updateChecklist = (newChecklist: IChecklist[]) => {
    checklist = newChecklist;
  };

  const { createItem, toggleItem, deleteItem, clearCompletedItems } = ManageChecklist({
    checklist,
    updateChecklist,
  });

  afterEach(() => {
    checklist = defaultChecklist;
  });

  it('create a new item in the checklist', () => {
    createItem('Spin round and round');
    expect(checklist).toMatchSnapshot();
  });

  it('complete an item', () => {
    const index = 0;
    const item = checklist[index];
    toggleItem({ item, index });
    expect(checklist).toMatchSnapshot();
  });

  it('delete an item', () => {
    deleteItem({ index: 2 });
    expect(checklist).toMatchSnapshot();
  });

  it('remove completed items from checklist', () => {
    clearCompletedItems();
    expect(checklist).toMatchSnapshot();
  });
});
