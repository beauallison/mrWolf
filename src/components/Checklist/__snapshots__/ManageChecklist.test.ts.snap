// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageChecklist complete an item 1`] = `
Array [
  Object {
    "complete": true,
    "name": "Live by harmless untruths",
  },
  Object {
    "complete": false,
    "name": "Research the foma",
  },
  Object {
    "complete": true,
    "name": "Busy busy busy",
  },
]
`;

exports[`ManageChecklist create a new item in the checklist 1`] = `
Array [
  Object {
    "complete": false,
    "name": "Spin round and round",
  },
  Object {
    "complete": false,
    "name": "Live by harmless untruths",
  },
  Object {
    "complete": false,
    "name": "Research the foma",
  },
  Object {
    "complete": true,
    "name": "Busy busy busy",
  },
]
`;

exports[`ManageChecklist delete an item 1`] = `
Array [
  Object {
    "complete": false,
    "name": "Live by harmless untruths",
  },
  Object {
    "complete": false,
    "name": "Research the foma",
  },
]
`;

exports[`ManageChecklist remove completed items from checklist 1`] = `
Array [
  Object {
    "complete": false,
    "name": "Live by harmless untruths",
  },
  Object {
    "complete": false,
    "name": "Research the foma",
  },
]
`;
